### DISCUSSION TITLE
Recipe failing to add content when using IRecipeMigrator

### DISCUSSION THREAD

**Discussion by jsalgado10** [OrchardCMS/OrchardCore#7575] [Q&A] [answered]
Hello

I am working on a new module that will add some new content type and parts, but I would also would like to create a new data content as well. I created a new recipe and are calling it using 

       public async Task<int> CreateAsync()
        {
            await recipeMigrator.ExecuteAsync("commentator.recipe.json", this);

            return 1;
        }

All the content types and part are getting created correctly, but the data content was not.

**Comment by jtkech** [comment]
Just did a PR #7582 to fix #7557 where @rjpowers10 creates content items in a migration recipe. The issue was that the items were not indexed, but at least he was able to create content items in a migration recipe.

So would need more info, is it when enabling your module through the admin OR on a fresh setup?

If on a fresh setup, in the main setup recipe, in the `feature` step section where all features to enable are specified, remove your module and then add it to another `feature` step section, and after all the things you may rely on. This is required because e.g. this is the `OC.Contents` feature that creates the `ContentItenIndex` table, so the 1st feature step need to complete before we can use this table in another recipe step (e.g. executed through a migration).

**Comment by jsalgado10** [comment, marked as answer]
I am enabling the feature through the admin UI. I have a migration recipe file that has a ContentDefinition section and a content section. The content definition works correctly, but the content never add the data.

**Reply by jtkech** [reply]
I'm working on this kind of things, just tried the same ;) it works but tried from a fresh setup

Have you any logged error?

Can you show us your recipe?

For your items, did you at least set `"Latest": true`, and for your type definition `"Listable": true`?

**Update**: Just tried with our menu module, by not enabling it in the blog recipe, by adding a menu item in the migration recipe of the menu module, then i did a fresh blog setup, then i enabled the menu module through the admin, and it was working.

**Reply by jsalgado10** [reply]
Here is my recipe
[commentator.recipe.zip](https://github.com/OrchardCMS/OrchardCore/files/5521904/commentator.recipe.zip)

Here is my Create Async method on the Migrations Class

 public async Task<int> CreateAsync()
        {
            await recipeMigrator.ExecuteAsync("commentator.recipe.json", this);

            return 1;
        }

**Reply by jtkech** [reply]
Also which OrchardCore version?

**After a quick look in your recipe i remembered that in my tests i first had an issue with `"[js: parameters('AdminUsername')]"`, the scripting module was failing to resolve it, so i just used hard coded strings e.g. `"Owner": "admin"`, can you try it?**

But in that case i had logged erros under app_data

**Update**:

Okay, for a setup we pass an object that have these parameters values

                await recipeExecutor.ExecuteAsync(executionId, context.Recipe, new
                {
                    context.SiteName,
                    context.AdminUsername,
                    context.AdminUserId,
                    context.AdminEmail,
                    context.AdminPassword,
                    context.DatabaseProvider,
                    context.DatabaseConnectionString,
                    context.DatabaseTablePrefix
                },
                _applicationLifetime.ApplicationStopping);

But not for a migration recipe where we only pass a `new object()`

            return await _recipeExecutor.ExecuteAsync(executionId, recipeDescriptor, new object(), ...);

**The `recipeMigrator.ExecuteAsync()` should also have an `object environment` parameter so that we could pass 
contextualized parameters**

**Reply by jtkech** [reply]
Okay, i think your content step fails because in a **migration** recipe there is no context *parameters*

So for testing try to replace this kind of lines

          "Owner": "[js: parameters('AdminUsername')]",
          "Author": "[js: parameters('AdminUsername')]",

By using hard coded strings

          "Owner": "admin",
          "Author": "admin",

**Reply by deanmarcussen** [reply]
**Note** On the most recent builds the `Owner` property must use the parameters `"Owner": "[js: parameters('AdminUserId')]"`

I need to look further at how that is dealt with in a recipe migration.

**Reply by deanmarcussen** [reply]
> The recipeMigrator.ExecuteAsync() should also have an object environment parameter so that we could pass
contextualized parameters

Agree @jtkech we should either be able to resolve and pass a `UserId` to the recipe migrators, or resolve it internally.

But as it is a migration, which may run from the admin, when a feature is enabled, or may run during site startup, when there is no logged in user, so no http context to get the current user from, it might have to be a query before the migration / as part of the migration.

Or maybe we need a js script that could resolve it.

So [js: siteOwner()]` - which is available on the `ISite` object.

Or on the import routine in the content manager we might force set the `Owner` property to the `SiteOwner` if it is set to null, or if the existing `Owner` is not found.  (requires query during import to then test and resolve it.)

Thinking...

**Reply by jsalgado10** [reply]
@jtkech I changed the recipe to have static values for the Owner and Author. That fixed the issue and the data was created successfully

**Reply by jtkech** [reply]
@jsalgado10 Cool, so did it answer to your question? If so please mark this discussion as answered, thanks

@deanmarcussen 

>  On the most recent builds the Owner property must use the parameters "Owner": "[js: parameters('AdminUserId')]"

~~So we need to also change `OC.Tenants.ApiController` and `OC.Tenants.SetupController` as they still use `AdminUsername`~~

Okay, i saw how `AdminUserId` is set by `_setupUserIdGenerator` in the setup service

> So [js: siteOwner()]- which is available on the ISite` object.

Didn't find it on the ISite` object.

@deanmarcussen can you file an issue for all of this? I could do it just to mention the missing environment parameter, but better that you also describe what we need around the admin user, new js scripts, ~~do we really need to update `OC.Tenants` controllers~~.

Then on my side i will think about how to retrieve contextual data e.g. when executed through a setup. When doing a setup, after a given feature is enabled through a step, this is in the scope of the following step that the shell is rebuilt and activated and then that the migration is called. So on setup a migration recipe is not an inner recipe but still a kind of child recipe for which we could retrieve contextual data, maybe through a dedicated httpContext feature (not a shell scope feature as run in a different one)

**Reply by deanmarcussen** [reply]
@jtkech I have not forgotten. Just getting notes together.
